pkg/types/container/v0.0.1/entry.go

```diff
@@ -0,0 +1,392 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Should be able to reconstruct the entry with the sig location and
+the container digest?
+
+*/
+
+package container
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/pki/x509"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/sigstore/sigstore/pkg/signature/payload"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/pkg/errors"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/pki"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/container"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := container.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	ContainerObj models.ContainerV001Schema
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() ([]string, error) {
+	var result []string
+
+	// TODO: Do we always expect x509 keys signing containers?
+	keyObj, err := x509.NewPublicKey(bytes.NewReader(v.ContainerObj.Signature.PublicKey.Content))
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := keyObj.CanonicalValue()
+	if err != nil {
+		log.Logger.Error(err)
+	} else {
+		keyHash := sha256.Sum256(key)
+		result = append(result, strings.ToLower(hex.EncodeToString(keyHash[:])))
+	}
+
+	result = append(result, keyObj.EmailAddresses()...)
+
+	// This is the hash of the signed payload.
```

Comment:
this is the hash over the marshalled JSON envelope, correct? do we have issues if the marshal/unmarshal of the signing envelope isn't consistently ordered?

Reply:
If I change it like below to be the SHA of the payload bytes I think we're still sort of good -- hoping that it is the SHA of the raw bytes returned by something like https://github.com/sigstore/cosign/blob/a7bd67c0a7314200ed72e5d9870911e6149d76bf/pkg/oci/internal/signature/layer.go#L59

(which is currently what hashed rekord does -- I don't know if this totally "matches spec")